src/Storages/PostgreSQL/PostgreSQLReplicationHandler.cpp

```diff
@@ -0,0 +1,458 @@
+}
+
+
+void PostgreSQLReplicationHandler::waitConnectionAndStart()
+{
+    try
+    {
+        /// Will throw pqxx::broken_connection if no connection at the moment
+        connection->conn();
+
+        startSynchronization();
+    }
+    catch (const pqxx::broken_connection & pqxx_error)
+    {
+        LOG_ERROR(log, "Unable to set up connection. Reconnection attempt will continue. Error message: {}",
+                pqxx_error.what());
+
+        startup_task->scheduleAfter(reschedule_ms);
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+
+void PostgreSQLReplicationHandler::shutdown()
+{
+    stop_synchronization.store(true);
+    consumer_task->deactivate();
+}
+
+
+void PostgreSQLReplicationHandler::startSynchronization()
+{
+    createPublicationIfNeeded(connection->conn());
+
+    auto replication_connection = std::make_shared<PostgreSQLConnection>(fmt::format("{} replication=database", connection->conn_str()));
+    replication_connection->conn()->set_variable("default_transaction_isolation", "'repeatable read'");
+    auto ntx = std::make_shared<pqxx::nontransaction>(*replication_connection->conn());
+
+    std::string snapshot_name, start_lsn;
+
+    auto initial_sync = [&]()
+    {
+        createReplicationSlot(ntx, start_lsn, snapshot_name);
+        loadFromSnapshot(snapshot_name, storages);
+    };
+
+    /// Replication slot should be deleted with drop table only and created only once, reused after detach.
+    if (!isReplicationSlotExist(ntx, replication_slot))
+    {
+        initial_sync();
+    }
+    else if (!Poco::File(metadata_path).exists() || new_publication_created)
+    {
+        /// In case of some failure, the following cases are possible (since publication and replication slot are reused):
+        /// 1. If replication slot exists and metadata file (where last synced version is written) does not exist, it is not ok.
+        /// 2. If created a new publication and replication slot existed before it was created, it is not ok.
+        dropReplicationSlot(ntx);
+        initial_sync();
+    }
+    else
+    {
+        for (const auto & [table_name, storage] : storages)
+        {
+            try
+            {
+                nested_storages[table_name] = storage->getNested();
+                storage->setNestedLoaded();
+            }
+            catch (...)
+            {
+                tryLogCurrentException(__PRETTY_FUNCTION__);
+            }
+        }
+    }
+
+    ntx->commit();
+
+    consumer = std::make_shared<PostgreSQLReplicaConsumer>(
+            context,
+            connection,
+            replication_slot,
+            publication_name,
+            metadata_path,
+            start_lsn,
+            max_block_size,
+            allow_minimal_ddl,
+            is_postgresql_replica_database_engine,
+            nested_storages);
+
+    consumer_task->activateAndSchedule();
+
+    replication_connection->conn()->close();
```

Comment:
We need RAII-wrapper if connection is not closed automatically in destructor

Reply:
It is closed automatically. Removed.